core/modules/qdisp/QueryRequest.cc

```diff
@@ -213,6 +252,15 @@ bool QueryRequest::isCancelled() {
 }
 
 void QueryRequest::cleanup() {
+    LOGF_DEBUG("%1% QueryRequest::cleanup()" %_jobId);
+    {
+        std::lock_guard<std::mutex> lock(_finishStatusMutex);
+        if (_finishStatus == ACTIVE) {
+            LOGF_ERROR("%1% QueryRequest::cleanup called before finished" % _jobId);
+            return;
+        }
+    }
+    // These need to be outside the mutex lock, or you could delete the mutex before it is freed.
```

Comment:
I think this should work. It depends on only being called by _finish or _errorFinish, so I added some comments to that effect.
```
if (_keepAlive != nullptr) {
        _jobQuery.reset();
        _keepAlive.reset();
} else {
        _jobQuery.reset();
}
```